### ISSUE TITLE
Variable uniquifier

### ISSUE THREAD

**Issue by PeterTh** [insieme/insieme#5]
Create a pass which makes variable IDs unique. (probably extend normalizer)
For debugging purposes.

**Comment by HerbertJordan** [comment]
Can you provide a use-case for this request?

I see two issues with this:
- code grow (potentially exponential) => reduced scalability
- invalid assumption of unique variables leading to more subtle errors in analysis and tools

The best such a utility could do is to make variables 'unique' in the sense variable addresses are 'unique'. If you have this requirement, you may want to use variable-addresses instead.

However, for accurate analysis you anyway have to find a more suiting representation for variables fitting your abstraction of e.g. context information.

Recommendation: I guess what is actually required is something that is, based on two variable addresses, determining whether those are referring to the same variable or not. Solving this by making them pseudo-unique is probably not the right approach.

**Comment by 0ife** [comment]
Notes:
* This feature requested by @klois .
* The reason why I want to work at this is not to check whether two variables are referring to the same memory address but I want to increase readability when dumping IR. Reusing the variable name in different contexts makes it very hard to follow the output.
* Resulting from the above point, the uniquifying of variables is purely optional, and hence
  * invalid assumption of unique variables leading to more subtle errors in analysis and tools ➤ is no issue
  * code grow (potentially exponential) => reduced scalability ➤ may be an issue, but is user-induced (the one using the variable uniquifier) so it should be no big concern in general

**Comment by HerbertJordan** [comment]
Can't we make this an option of the pretty printer which is mapping variables dynamically to a list of fresh variables as soon as their definition is encountered? Note that every definition is always encountered before a variable's usage.

**Comment by 0ife** [comment]
@HerbertJordan: Could you please elaborate on [your earlier comment](https://github.com/insieme/insieme/issues/29#issuecomment-90883190)?
@klois and I find it useful to have unique variable IDs **in some cases** for analysis and transformation, so the possibility of implementing the feature in the PrettyPrinter has some drawbacks. I am currently doing the work as a frontend extension/plugin (so that we can manually enable it for a single AST tree if needed) but it would be nice if we agreed upon a long-term roadmap reg variable representation before a lot of implementation effort will be thrown out again.

**Comment by HerbertJordan** [comment]
My 50 cents on variable IDs: It is complicated and expensive to make their IDs unique => thus dealing with ID collisions within analysis / IR tools is just necessary. And, as already mentioned earlier, for some cases involving closures, dynamic dispatching and/or recursions different abstractions for variables than static IR variables have to be found anyway. Exposing unique IDs can at most be as expressive as utilizing variable addresses instead of variable pointers.

@0ife regarding this comment: back in 2011, when i spend a lovely week at europar, I had the idea of a variable-id free IR. Actually, I think I was challenged to demonstrate that something like that could exist. Each expression would essentially define a list of free and bound variables which would be recursively linked with those of nested expressions. For instance, a constant
```C
1
```
has no free / bound variables. A variable
```C
a
```
would have one free variable (itself), no bound variables. It would be represented by a variable node without any id, just stating the fact that it has a free variable. The same is true for
```C
b
```
such that essentially both variables are represented by the same node. The distinction comes with the composition. If you have something like
```C
a + b
```
you get a call node with 2 free variables, 0 bound variables. The first free variable is linked to the first free variable of the first child, the second free variable to the first free variable of the second child. This fully characterizes this term and all its deducible properties. The same structure covers
```C
c+d
```
as well but not
```C
x+x
```
or
```C
bla+bla
```
which would be both covered by a call node with only one free variable linked up to the first free variable of its two sub-nodes. Extend this thought to the entire IR.

The result is that the naming of variables is no longer a distinguishing element of code fragments. It is like getting rid of the alpha-conversion in lambda calculus by moving it into the structure. I even wrote a prototype of this and it is still there
https://github.com/insieme/insieme/blob/master/code/core/test/no_variable_name_test.cc

The down side is that all variables in the IR are represented by one unified node. Something like name-annotations (which back then didn't exist and are actually still just 'decoration') would not work any more. Also, it is hard to say how difficult it would be to work with such a representation. For instance, in utilities like the pretty printer those names would have to be implicitly synthesized to enable human readers to understand IR constructs. 

Any thoughts?

**Comment by klois** [comment]
What would be the advantage (despite proving that it's doable)?
Furthermore, did I get you wrong or would it then impossible to distinguish between a+b and c+d?

**Comment by klois** [comment]
VariablePtrs with unique Ids may not be more expressive than VariableAddresses (obviously), but they are more convenient to use in many cases. For example, if you want to compare it two expressions are referring to the same variable, with VariablePtrs with non-unique Ids one has to check the id and the scope (which is not possible if you only have the Ptr). With VariableAddresses the only way I found is to go up the AST to look for the declaration and check if that is the same for both Variable Addresses. With unique Ids one would only have to comapre the Ids...

**Comment by HerbertJordan** [comment]
The advantage: no variable names, no risk of accidental unwanted capturing of free variables due to name collisions when conducting re-writing enabling more localized code transformations potentially leading to much faster operations; being closer to a (impossible) normal form for code. A more compact representation.

And yes, a+b and c+d would, without a context, be distinguishable.  So is the free lambda calculus term (a b) from (c d). If I'd write it like it as a function
```C
f(a,b) { return a + b; }
```
and 
```C
f(c,d) { return c + d; }
```
nobody would ask for those to be distinguishable.

**Comment by klois** [comment]
Except for one who wants to perform some operation on a but not on c. Imagine you want to exchange a with a variable of a different type but keep c. Wouldn't that be hard in a no-names representation?

**Comment by HerbertJordan** [comment]
@0ife 
>With VariableAddresses the only way I found is to go up the AST to look for the declaration and check if that is the same for both Variable Addresses. With unique Ids one would only have to comapre the Ids...

Yes, right. That's the way to go. What you would do with your make-variable-unique-tool would be simply to apply this on every variable as a preemptive operation just in case you want to check for this property later -- with the additional responsibility of preserving this property. When at some point you are then actually modifying the IR you will have to scan through the entire code fragment to make sure that a new variable id you are about to utilize is indeed not yet used. If you ask me, just offering a comparison tool for VariableAddresses (which I'm sure is somewhere around already) sounds much more stable.

**Comment by klois** [comment]
If we have it, please show me

**Comment by HerbertJordan** [comment]
@klois 

> Except for one who wants to perform some operation on a but not on c. Imagine you want to exchange a with a variable of a different type but keep c. Wouldn't that be hard in a no-names representation?

First: it is just an idea and ultimately only a firm investigation will cover all details. But you would definitely have to say goodby to the concept of "variable a" or "variable c". You'd address them as the first free variable in expression X. And if you want to exchange that variable you'd have to follow all the links starting from your root node to where it is referenced and replace its usage. This is like build-int DEF-USE chance for symbol names.

**Comment by klois** [comment]
I agree. The problem is that you have to do the DEF-USE analysis every time you do anything as you are not able any more to "remember" a certain construct. That would probably be a performance issue.

**Comment by HerbertJordan** [comment]
I have an implementation here:
https://github.com/insieme/insieme/blob/master/code/analysis/include/insieme/analysis/cba/utils/cba_utils.h#L59

It is based on VariableInstances which is a more generalized Version of VariableAddresses. But you can take it as a template and provide add a similar function to 
https://github.com/insieme/insieme/blob/master/code/core/include/insieme/core/analysis/ir_utils.h#L376
where it should be.

**Comment by HerbertJordan** [comment]
The DEF-USE analysis results are coded up directly in the IR nodes in this model. It is the central construct and produced as a side-effect of the IR construction.

**Comment by 0ife** [comment]
Done with commit 1f5e9a2bf4723e02aa08050603949c020c5e4963

**Comment by HerbertJordan** [comment]
Can you briefly describe which approach you took, to conclude the discussion here?

**Comment by 0ife** [comment]
After several complete rewrites I found the following approach to be most sensible:
* Through a visitor, visit all the variables in the code (which is given as NodeAddress, so the variable replacements are confined to a particular IR tree).
* Build up a map that maps a VariablePtr to a vector of VariableAddresses. This vector contains all distinct variables in the code, as found through the visitor.
* Count the items in the map which corresponds to the number of unique variables found in the code. Number all of these variables in a consecutive way. The number that the set of variables is assigned will be the target variable number.
* Perform the variable renaming in an iterative way with the help of replaceAll; in each step rename only one variable. In case the target variable name is still taken, do a rename to the lowest non-taken variable ID and reschedule the rename till the end. Proceed with the next variable, until the list of replacements is empty. Each time, do the replacement on the tree that was returned by the previous call to replaceAll.

I acknowledge that this is most probably not the fastest algorithm, but it is rather short, easy to understand and easy to maintain. Speed is not an issue (until now), thus so far no need to optimize.

**Comment by HerbertJordan** [comment]
thx
